In [4]:
import pandas as pd
import sklearn
import sys
import plotly.express as px
import warnings
warnings.filterwarnings('ignore')

from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor

In [12]:
sys.path.append('src/')

import plot as p
import utils as u
import model_experiments as exp

pd.set_option('display.max_columns', 500)

In [13]:
# Load data 
df = pd.read_csv('training_data/kc_house_data.csv') 

In [14]:
# dropping duplicate rows keeping latest ones
df.sort_values(['id', 'price']).drop_duplicates(subset = 'id', keep = 'last', inplace = True)

## Modelling experiments

In [15]:
# get feature and target dataframes
feat_df = df[df.loc[ :,df.columns != 'price'].columns]
target_df = df['price']

# get train_test_split dfs
feat_train, feat_test, target_train, target_test = u.get_train_test_split(
                                                            feat_df,
                                                            target_df,
                                                            random_state = 1
                                                  )

In [16]:
feature_sets_dict = { 
                      'all_preprocessed_features': feat_df.columns,
                      'high_correlated_features': [
                                                    'date', 'id', 'sqft_basement', 'sqft_above', 'lat', 
                                                    'long', 'yr_built', 'yr_renovated', 'zipcode', 'sqft_lot',
                                                    'sqft_living15', 'sqft_lot15', 'floors',
                                                    'bedrooms', 'sqft_living', 'grade',
                                                    'bathrooms', 'view', 'waterfront'
                                                   ]
                    }

# list to store cross validation metrics from all experiments
metrics = []

In [7]:
exp.run_model_experiments?

### Linear Regression

In [17]:
lm = LinearRegression()

metrics = exp.run_model_experiments(
                    lm,
                    'linear_regression',
                    feat_train,
                    target_train,
                    feature_sets_dict,
                    metrics
)

[TRAINING] linear_regression ------
[EXP 0] all_preprocessed_features [DONE]
[TRAINING] linear_regression ------
[EXP 1] high_correlated_features [DONE]


### Ridge regression

In [18]:
# run model experiments on ridge regression model
RidgeModel = Ridge(alpha = 0.1)
metrics = exp.run_model_experiments(
                    RidgeModel,
                    'ridge_regression',
                    feat_train,
                    target_train,
                    feature_sets_dict,
                    metrics
)

# get fitted grid search object
grid = u.get_fitted_grid_object(
                        RidgeModel,
                        [{'model__alpha': [0.001,0.1,1, 10, 100, 1000, 10000]}],
                        feat_train,
                        target_train
                )
            
best_params_dict = {
                     k.split('__')[1] : v
                         for k,v in grid.best_params_.items()
                   }

# run model experiments with grid search best parameters on ridge model
RidgeModel_grid = Ridge(**best_params_dict)
metrics = exp.run_model_experiments(
                    RidgeModel_grid,
                    'ridge_regression_grid',
                    feat_train,
                    target_train,
                    feature_sets_dict,
                    metrics
)

[TRAINING] ridge_regression ------
[EXP 0] all_preprocessed_features [DONE]
[TRAINING] ridge_regression ------
[EXP 1] high_correlated_features [DONE]
[Running Gridsearch]
[TRAINING] ridge_regression_grid ------
[EXP 0] all_preprocessed_features [DONE]
[TRAINING] ridge_regression_grid ------
[EXP 1] high_correlated_features [DONE]


In [9]:
best_params_dict

{'alpha': 100}

### Random Forest Regressior

In [19]:
# run model experiments for random forest
RF = RandomForestRegressor(n_estimators=150, n_jobs=2, max_features=3, max_depth=3, random_state=1)
metrics = exp.run_model_experiments(
                    RF,
                    'random_forest_regressor',
                    feat_train,
                    target_train,
                    feature_sets_dict,
                    metrics
)

params = [
            {
                'model__n_estimators': [150, 200],
                'model__max_features': [8, 10, 20],
                'model__max_depth': [10, 15, 20]
            }
         ]

# fit grid search object
grid = u.get_fitted_grid_object(
                    RF,
                    params,
                    feat_train,
                    target_train
                    )
          
best_params_dict = {
                     k.split('__')[1] : v
                         for k,v in grid.best_params_.items()
                   }

# run model experiments for random forest with grid search best params
RF_grid = RandomForestRegressor(**best_params_dict, random_state=1)
metrics = exp.run_model_experiments(
                    RF_grid,
                    'random_forest_regressor_grid',
                    feat_train,
                    target_train,
                    feature_sets_dict,
                    metrics
)

[TRAINING] random_forest_regressor ------
[EXP 0] all_preprocessed_features [DONE]
[TRAINING] random_forest_regressor ------
[EXP 1] high_correlated_features [DONE]
[Running Gridsearch]
[TRAINING] random_forest_regressor_grid ------
[EXP 0] all_preprocessed_features [DONE]
[TRAINING] random_forest_regressor_grid ------
[EXP 1] high_correlated_features [DONE]


In [22]:
best_params_dict

{'max_depth': 15, 'max_features': 10, 'n_estimators': 200}

### Gradient Boosting Regressor

In [21]:
# run model experiments for gradient boosting 
gb = GradientBoostingRegressor(random_state=1)
metrics = exp.run_model_experiments(
                    gb,
                    'gradient_boosting_regressor',
                    feat_train,
                    target_train,
                    feature_sets_dict,
                    metrics
)

params = [
            {
                'model__n_estimators': [100, 150],
                'model__max_features': [12, 20],
                'model__max_depth': [12, 20]
            }
         ]

# fit grid search object
grid = u.get_fitted_grid_object(
                        gb,
                        params,
                        feat_train,
                        target_train
                )
            
best_params_dict = {
                     k.split('__')[1] : v
                         for k,v in grid.best_params_.items()
                   }

# run model experiments for random forest with grid search best params
gb_grid = GradientBoostingRegressor(**best_params_dict, random_state=1)
metrics = exp.run_model_experiments(
                    RF_grid,
                    'gradient_boosting_regressor_grid',
                    feat_train,
                    target_train,
                    feature_sets_dict,
                    metrics
)

[TRAINING] gradient_boosting_regressor ------
[EXP 0] all_preprocessed_features [DONE]
[TRAINING] gradient_boosting_regressor ------
[EXP 1] high_correlated_features [DONE]
[Running Gridsearch]
[TRAINING] gradient_boosting_regressor_grid ------
[EXP 0] all_preprocessed_features [DONE]
[TRAINING] gradient_boosting_regressor_grid ------
[EXP 1] high_correlated_features [DONE]


In [22]:
best_params_dict

{'max_depth': 12, 'max_features': 12, 'n_estimators': 150}

### Polynomial transformation

In [23]:
# Fit and transform the variables with 2nd order polynomial and run exp on linear regression
poly_lm = LinearRegression()
metrics = exp.run_model_experiments(
                    poly_lm,
                    'poly_linear_regression',
                    feat_train,
                    target_train,
                    feature_sets_dict,
                    metrics,
                    is_poly=True
)

# Fit and transform the variables with 2nd order polynomial and run exp on ridge regression
poly_ridge = Ridge(alpha = 0.01)
metrics = exp.run_model_experiments(
                    poly_ridge,
                    'poly_ridge_regression',
                    feat_train,
                    target_train,
                    feature_sets_dict,
                    metrics,
                    is_poly=True
)

[TRAINING] poly_linear_regression ------
[EXP 0] all_preprocessed_features [DONE]
[TRAINING] poly_linear_regression ------
[EXP 1] high_correlated_features [DONE]
[TRAINING] poly_ridge_regression ------
[EXP 0] all_preprocessed_features [DONE]
[TRAINING] poly_ridge_regression ------
[EXP 1] high_correlated_features [DONE]


In [24]:
metrics_df = pd.DataFrame(metrics).sort_values(by='test_mae_mean', ascending = True)
metrics_df

model_name                feature_set  \
8       random_forest_regressor_grid  all_preprocessed_features   
12  gradient_boosting_regressor_grid  all_preprocessed_features   
13  gradient_boosting_regressor_grid   high_correlated_features   
9       random_forest_regressor_grid   high_correlated_features   
10       gradient_boosting_regressor  all_preprocessed_features   
11       gradient_boosting_regressor   high_correlated_features   
4              ridge_regression_grid  all_preprocessed_features   
2                   ridge_regression  all_preprocessed_features   
0                  linear_regression  all_preprocessed_features   
5              ridge_regression_grid   high_correlated_features   
3                   ridge_regression   high_correlated_features   
1                  linear_regression   high_correlated_features   
16             poly_ridge_regression  all_preprocessed_features   
17             poly_ridge_regression   high_correlated_features   
7            random_forest_regressor   high_correlated_features   
6            random_forest_regressor  all_preprocessed_features   
15            poly_linear_regression   high_correlated_features   
14            poly_linear_regression  all_preprocessed_features   

    train_mse_mean  test_mse_mean  train_mse_std  test_mse_std  \
8     61559.385207   1.450462e+05    9432.659907  3.306299e+04   
12    61559.385207   1.450462e+05    9432.659907  3.306299e+04   
13    61416.146621   1.455435e+05    8820.489380  3.210081e+04   
9     61416.146621   1.455435e+05    8820.489380  3.210081e+04   
10   118893.106552   1.398531e+05   21243.697036  2.874415e+04   
11   118812.405336   1.417188e+05   20630.498047  2.853717e+04   
4    163251.008832   1.686620e+05   37995.881416  8.184802e+04   
2    164238.267539   1.696817e+05   38684.157776  8.419962e+04   
0    164677.560694   1.714312e+05   37319.196303  9.054877e+04   
5    166267.340683   1.721514e+05   40559.016816  8.769053e+04   
3    167310.104081   1.732306e+05   41282.494590  9.012705e+04   
1    167478.178988   1.733657e+05   41557.370360  9.047650e+04   
16   124698.628288   1.516860e+05   15500.254419  4.791633e+04   
17   126409.529096   1.536363e+05   15959.955966  4.857935e+04   
7    269682.931687   2.694136e+05   33662.925300  7.903991e+04   
6    272887.519699   2.731380e+05   36596.567927  7.705030e+04   
15   126409.895863   1.797723e+13   15960.406374  2.542364e+13   
14   124730.918330   7.849986e+14   15183.655804  1.110095e+15   

    train_mae_mean  test_mae_mean  train_mae_std  test_mae_std  train_r2_mean  \
8     36832.465863   8.824707e+04     437.576104  2.355828e+03       0.969755   
12    36832.465863   8.824707e+04     437.576104  2.355828e+03       0.969755   
13    36972.112762   8.883986e+04     339.298819  1.887229e+03       0.969893   
9     36972.112762   8.883986e+04     339.298819  1.887229e+03       0.969893   
10    71765.909782   8.926845e+04     916.912488  2.438414e+03       0.887175   
11    72001.241436   9.116246e+04     848.704054  2.648621e+03       0.887329   
4     94474.472579   9.719014e+04     326.394665  1.020057e+03       0.787267   
2     94532.531831   9.724125e+04     326.281419  1.145849e+03       0.784686   
0     94604.897623   9.736940e+04     263.628717  1.316984e+03       0.783516   
5     95565.827660   9.847164e+04     387.772420  1.122012e+03       0.779339   
3     95618.391318   9.852670e+04     389.298935  1.263983e+03       0.776562   
1     95670.991390   9.855853e+04     432.687689  1.242906e+03       0.776115   
16    81511.230163   1.017882e+05     467.993583  3.402023e+03       0.875884   
17    82595.934483   1.027505e+05     440.979898  3.431250e+03       0.872455   
7    152214.795795   1.534682e+05     702.537844  2.651798e+03       0.419537   
6    155237.288316   1.567887e+05     872.455104  2.399591e+03       0.405673   
15    82595.280576   1.367180e+11     441.151031  2.734358e+11       0.872454   
14    81528.706661   6.025323e+12     4

## Plot evaluation metrics of experiments

In [31]:
x = metrics_df[~(metrics_df['model_name']=='poly_linear_regression')]
fig = px.line(x, x="model_name", y="test_mae_mean", title='MAE vs Model experiment', color = 'feature_set')
fig.show()

## Results --------

     ------- We looked at three evaluation metrics - RMSE - Root Mean Squared Error, MAE - Mean Absolute Error & R^2
             from 5-fold cross-validation. We intend to optimize for `MAE` & `R^2` as the metric considering the target
             variable has outliers and MAE is not sensitive to outliers.
     ------- The best model so far - `Gradient boosting regressor` with gridsearch parameters on all preprocessed 
             features leading to MAE on test set - 99872 & R^2 on test set - 79%.